Admin panel

Kaszana102 opened the pull request.

Kaszana102 committed:
creating admin panel

created files:
-functional test admin_panel_test.py
-web files:  admi_panel.css, admin_panel.html
file modification:
- README.md - adding env variables with bash described
- database.py - custom class for mail which queries CSVDatabase
- invoice_generator.py - modfied single dictionary entry for compatibility with user_dto.py
-mail.py - simple fix
-main.py - addedd new site /admin_panel, which reacts with GET params
-main.css  - added import
-index.html - added link to admin_panel - temporary solution, because logging in isn't implemented
-text_generator.py - addedd type hint
-user_dto.py - addedd field: subscriptions. It allows checking which subscriptions are subscribed
-invoice_generator_test.py - dictionary key fix
-test_text_generator.py - fix test

Kaszana102 committed:
Merge branch 'main' into admin-panel

Kaszana102 committed:
Addedd pdf sending button

file modifications:
- csvDatabase.py - addedd checking if values is a list (for example for checking packets)
- database.py - remove checking if list, because it was moved to csvDatabase.py
- invoice_generator.py - dictionary key change. Compatibility with user_dto
- mail.py - small fix
- text_generator.py - changed text, to match invoice
- user_dto.py - fixed compatibility with invoice

Kaszana102 committed:
generating mail text fix

Kaszana102 committed:
addedd href to index in admin_panel

Kaszana102 committed:
Merge branch 'main' into admin-panel

Kaszana102 committed:
Lint html fix

Kaszana102 committed:
kebab case id

Kaszana102 committed:
Coverage fix

coverage fix!

Kaszana102 committed:
Merge branch 'main' into admin-panel

Kaszana102 committed:
Total fix

-reverted some changes (mostly lists in csv)
-Database now contains DatabaseContext
-modified invoice to work with ClientInfo
-fixed unit tests accordingly
-addet data_set file for quick generating values for tests

Kaszana102 committed:
Lint test

Kaszana102 committed:
function test test

Kaszana102 committed:
func test

Kaszana102 committed:
func test

Kaszana102 committed:
deleted invoice xml from test

MikolekN commented on test/user_dto_test.py:
Unnecessary change.

Kaszana102 replied:
i generating text was modified, i had also to modify the tests :/

MikolekN replied:
These tests are unnecessary because they duplicate the code of other tests:
test_xml_serialization_deserialization_empty_packets is the same as test_xml_serialization_deserialization
test_csv_serialization_deserialization_empty_packets is the same as test_csv_serialization_deserialization

Kaszana102 replied:
oh wait, i looked at the wrong file. Ok right it's useless. I might not have  reverted oll useless changes. Tests removed

MikolekN commented on src/user_dto.py:
Unnecessary change.

Kaszana102 replied:
actually some of it was needed. Firstly at the beginning i hadn't fully understood how to do we treat bundles and subscription, so I changed the text accordingly. Payment method is nowhere stored, so i removed if form the text.

MikolekN replied:
The additional text in the comment isn't necessary because it doesn't accurately reflect the code.

Kaszana102 replied:
jeez, i looked at the file above the comment, not below. ( i looked at generator_text.py)
The comment is also a part of history.  Removed.

MikolekN commented on functional/functional_test_constant.py:
What is the need for this file?

Kaszana102 replied:
I was told to the test for the admin panel. So they only check if button works, to check the notificiation of clicked button. There is no way to test sending mail, and invoice this way.

MikolekN replied:
What I meant is that from what I can see the file is only used in admin_panel_test.py where it's used to set a value for self.BASE_URL. Why not set it the same way as in other tests?
Instead of: self.BASE_URL = BASE_URL + '/admin_panel'
Use: self.BASE_URL = os.getenv("BASE_URL")
And then:
Instead of: self.driver.get(self.BASE_URL)
Use: self.driver.get((os.path.join(self.BASE_URL, "admin_panel")))

Kaszana102 replied:
@BognaLew told me to create some file with BASE_URL, instead of geting url form env.

MikolekN replied:
ok

BognaLew replied:
I think you misunderstood me. You should be using environmental variable, but you need to remember to set it in PyCharm in run configuration. Please switch to environmental variable before merge.

Kaszana102 replied:
switched

Kaszana102 committed:
Removed copied tests

Kaszana102 committed:
removed comment

Lenetis commented on src/csvDatabase.py:
```diff
@@ -71,10 +71,15 @@ def add_client(self, client_data: Dict) -> None:
-            writer.writerow(client_data)
+            for key in client_data_copy.keys():
+                if type(client_data_copy[key]) is list:
+                    client_data_copy[key] = '['+'.'.join(client_data_copy[key])+']'
```
1. What if `type(client_data_copy[key])` is a `tuple` and not a `list`?
2. Why convert lists to string in a manual way, instead of just `str(list)`?
3. Why is this list separated by dots and not commas?

Kaszana102 replied:
0. removed lists in csv (Bartek made managing packets in a more sensible way)
2. because commas
3. because commas (csv was comma separated)

Lenetis commented on src/csvDatabase.py:
```diff
@@ -86,6 +91,14 @@ def get_client(self, client_id: str) -> Dict | None:
+                    #check for lists values
+                    for key in data:
+                        if '[' in data[key]:
+                            data[key] = data[key].replace('[', '').replace(']', '')
```
Why convert string to list in a manual way?
If you converted lists to strings with just `string = str(list)`, you could use `list = ast.literal_eval(string)` here.

Kaszana102 replied:
removed funcionalisty

Lenetis commented on src/csvDatabase.py:
```diff
@@ -150,6 +163,14 @@ def get_clients(self, predicate: Callable[[Dict], bool] = None) -> List:
+                    # check for lists values
+                    for key in data:
+                        if '[' in data[key]:
+                            data[key] = data[key].replace('[', '').replace(']', '')
```
Manual conversion of string to list, same comment as before.

Kaszana102 replied:
removed lists in csv functionality

Lenetis commented on src/database.py:
```diff
@@ -14,33 +20,63 @@ def get_user_sex(self, user_id) -> str:
+        if client['gender'] == 'female':
+            return 'F'
+        else:
+            return 'M'
```
The registration form also has "Other" gender

Kaszana102 replied:
how i am supposed to write mail to nonbinary person?

Lenetis commented on src/database.py:
```diff
@@ -14,33 +20,63 @@ def get_user_sex(self, user_id) -> str:
+        subscribed_packets_names = [x.name for x in subscribed]
 
-        return []
+        for service in PACKETS.keys():
```
Little nitpick, but why "for **service** in PACKETS.keys()"? Woudn't "for **packet** in PACKETS.keys()" be more fitting?

Kaszana102 replied:
renames

Lenetis commented on src/database.py:
```diff
@@ -14,33 +20,63 @@ def get_user_sex(self, user_id) -> str:
+    def get_subscription(self, user_id) -> str:
+        """
+        Function which retrieves all services subscribed by user
+        :param self:
```
`:param self` is unnecessary in a docstring.
`user_id` should have a typehint.

Lenetis commented on src/mail.py:
```diff
@@ -31,22 +31,22 @@ def send_mail(receiver: str, subject: str, content: str, files: List[str]) -> No
             )
         # After the file is closed
-        part['Content-Disposition'] = 'attachment; filename="{basename(f)}"'
+        part['Content-Disposition'] = 'attachment; filename="%s"' % basename(filename)
```
FYI, the previous f-string didn't work because it lacked an `f` in front.
`f'attachment; filename="{basename(f)}"'` would work fine.

Lenetis commented on src/main.py:
```diff
@@ -1,9 +1,21 @@
+
+import inspect
+from dataclasses import asdict
+import os
```
- Wrong import order (imports within one grup should be in alphabetic order, and `import ...` should be before `from ... import`)
- Unnecessary newline after datetime import
- inspect and asdict import statements are unused. Also generate_invoice_xml and process_form which are in the next few lines

https://peps.python.org/pep-0008/#imports

Lenetis commented on src/main.py:
```diff
@@ -156,3 +172,37 @@ def edit_subscription():
+                client = db.get_client_by_email(clients[0]['email'])
+                file_name = 'invoice.pdf'
+                invoice = Invoice(client)
+                invoice.save_pdf('invoice.pdf')
```
In line 194 you define a variable called "`file_name`", but here the filename is given as a string

Lenetis commented on src/templates/admin_panel.html:
```diff
@@ -0,0 +1,39 @@
+	<div class="horizontal-flex">
+		<div class="admin-button">
+			<input type="submit" value="suggest" name="suggest-services" id="suggest-services">
+			<label for="suggest-services" class="vertical-flex" style="height: 100%;" id="suggest-services-label">
```
Why use `style="..."` in html instead of using css?

Lenetis commented on src/templates/admin_panel.html:
```diff
@@ -0,0 +1,39 @@
+{% block content %}
+
+<h1>DC Paint Drying</h1>
+<a href="..">Main Page</a>
```
In `index.html` you use `<a href="{{ url_for('admin_panel') }}">`. Could a similar method be used here, instead of the ".."?

Lenetis commented on src/text_generator.py:
```diff
@@ -1,7 +1,9 @@
+from src import manifest
```
Unused import statement

Kaszana102 replied:
removed line

Lenetis commented on src/user_dto.py:
```diff
@@ -70,4 +70,5 @@ def from_csv(cls, csv_str):
         for i in range(len(data)):
             if data[i].isdigit():
                 data[i] = int(data[i])
+
```
Unnecessary newline

Kaszana102 replied:
literaly pycharm sees it as a problem, when there is no newline there XD

Lenetis replied:
Ok, in that case I retract my comment. I joust found it strange that the only change you made to this file was adding a newline, and I thought it was a mistake

Lenetis commented on test/csv_test.py:
```diff
@@ -174,3 +174,45 @@ def test_use_existing_csv_file_with_auto_columns():
+    assert db2.get_client("01") == client
+
+    db1.drop_database()
+def test_get_all_clients_with_list():
```
Rename to `test_get_all_clients_with_packets_list` (other tests you added follow this convention)

Kaszana102 replied:
removed tests

Lenetis commented on test/database_test.py:
```diff
@@ -0,0 +1,55 @@
+from src.database import Database
+from src.database_context import DatabaseContext
+from src.subscription_info import SubscriptionInfo
+from src.user_dto import UserDto
```
Most of these import statements are unused.

Lenetis commented on test/test_text_generator.py:
```diff
@@ -1,102 +1,118 @@
+        self.db.get_user_surname = MagicMock(return_value='Wierzba')
 
-        assert text_gen.replace_keywords('{$greeting}', 1, database) == 'Szanowny Panie Wierzba'
+        assert text_gen.replace_keywords('{$greeting}', 1, self.db) == 'Szanowny Panie Wierzba'
```
`text_gen.replace_keywords('{$greeting}', 1, self.db)`
`Expected type 'str', got 'int' instead---^`

Same problem happens in other test

Kaszana102 replied:
changed

Kaszana102 committed:
removed lists in csv

-removed lists in csv
-fixed imports
-remoed tests depending on lists in csv

Kaszana102 committed:
coverage increase

Kaszana102 committed:
Coverage reduced by 1%

can't make tests for the checking mail functionality, in order not to spam our mailbox. They are supposed to be omitted.
Everything other is 100%.

Kaszana102 committed:
minor fixes
